content/concepts/peer-routing.md

```diff
@@ -3,6 +3,96 @@ title: Peer Routing
+
+For any given node, Kademlia divides the whole binary tree into a series of successively lower subtrees. The highest subtree consists of the half of the binary tree not containing the original node. The next subtree consists of the half of the remaining tree not containing the next highest node, and so on.
+
+You can see here for node `001` there are three levels of subtrees are also circle, consisting of all nodes with prefixes `1` (highest level subtree) , `01`, `000` (lowest level subtree).
+
+The Kademlia protocol ensures that every node knows of at least one node in each of its subtrees, if that subtree contains a node. This guarantees that any node can be located by it's peer ID.
+
+Let's see how once can locate in an example. We will call the red node 'Bob', with the prefix `001.` We will call the yellow node Alice, with the prefix `01`
+
+_A brief reminder_: Alice's peerID is a key in the DHT, and the value of this key is Alice's IP address.
+
+Bob can locate Alice in this example. Bob knows Alice's peerID and he needs to get her IP address. From the subtree with the prefix `1` Bob knows a peer with prefix `101`. He sends a message to their address, where he asks for Alice's address.
+
+![1-hop](1-hop.png)
+
+Peer `100` doesn't have Alice's key. But they have key of node `110` from their subtree which locates Alice better! And we are doing second hop:
+
+![2-hop](2-hop.png)
+
+Peer `110` also doesn't have Alice's key, but they again locate Alice better and send request to node with prefix `1111`.
+
+![3-hop](3-hop.png)
+
+Peer `1111` stores Alices peer id and can send the key to the Bob! This is how routing works and here is full scheme:
+
+![final-hop](final-hop.png)
+
+Peer `1111` sends response directly to Bob, since each request in this chain had Bob's address.
+
+### Routing table
+
+The place where one node stores info about other nodes is called a routing table. The routing table is a binary tree whose leaves are `k-buckets`.
+
+### k-buckets
+
+Take a look at the initial tree:
+![Kademlia-system](Kademlia-system.png)
+See this circled sub-trees? Remember that the Kademlia protocol ensures that every node knows of (and can locate) at least one node in each of its subtrees. If that subtree contains a node. `k-bucket` of a sub-tree is a set of that nodes. Once again: from each subtree we take only those nodes that we know - and we call it `k-bucket`. In out case `k-buckets` for Bob look like this
+
+![k-buckets](k-buckets.png)
+
+Each `k-bucket` covers some range of the ID space, and together the `k-buckets` cover the entire peer ID space with no overlap.
+
+## Dynamic DHT
+
+Okay, that was it for static DHT. And the gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.
```

Comment:
```suggestion
The gap from our current state to a fully dynamic DHT is not as far away as it seems. Let us first consider how we can extend this protocol to support computers sporadically leaving.
```

Reply:
Sorry for taking so much time 😅
And thanks!

Reply:
No worries at all, thanks for contributing and taking the time to write this!